unified-runtime/source/adapters/offload/context.hpp

```diff
@@ -17,12 +17,28 @@
 #include <unordered_map>
 #include <ur_api.h>
 
+struct alloc_info_t {
+  ol_alloc_type_t Type;
+  size_t Size;
+};
+
 struct ur_context_handle_t_ : RefCounted {
   ur_context_handle_t_(ur_device_handle_t hDevice) : Device{hDevice} {
     urDeviceRetain(Device);
   }
   ~ur_context_handle_t_() { urDeviceRelease(Device); }
 
   ur_device_handle_t Device;
-  std::unordered_map<void *, ol_alloc_type_t> AllocTypeMap;
+  std::unordered_map<void *, alloc_info_t> AllocTypeMap;
+
+  std::optional<alloc_info_t> getAllocType(const void *UsmPtr) {
+    for (auto &pair : AllocTypeMap) {
+      if (UsmPtr >= pair.first &&
+          reinterpret_cast<uintptr_t>(UsmPtr) <
+              reinterpret_cast<uintptr_t>(pair.first) + pair.second.Size) {
+        return pair.second;
+      }
+    }
+    return std::nullopt;
+  }
```

Comment:
I don't like that we have to track the allocations in both the adapter and in liboffload itself, it would be better if we could refactor that to only be in one place. Not something to be solved in this PR, just an idea for the future.

Reply:
Yeah that sounds good to me